keps/sig-cli/2227-kubectl-default-container/README.md

```diff
@@ -0,0 +1,300 @@
+
+### Proposal Details
+
+A single, generic and well known annotation for all above commands like `kubectl.kubernetes.io/default-container` is a good choice to avoid needing many new annotations in the future.
+
+There are currently 3 commands that consume this annotation if `--container` is not specified.
+- `kubectl exec`
+- `kubectl attach`
+- `kubectl cp`
+- `kubectl logs`
+
+However, there is an exception that `kubectl logs` will consume this annotation only if no `--all-containers` option is specified. If `--all-containers` is specified, all pods logs should be returned.
+
+`kubectl debug` will not consume this annotation, as `--container` for `kubectl debug` is to speicify new debug container name to use. It is quite differenet with the annotation meaning here.
+
+If `kubectl.kubernetes.io/default-logs-container` is specified, we should use this annotation instead of the general one for `kubectl log` and use general annotation for other commands.
+
+### Notes/Constraints/Caveats
+
+As the annotation `kubectl.kubernetes.io/default-container` will not be automatically added, users and Pod owners will need to be aware of this new annotation.
+
+### Risks and Mitigations
+**Note:** No server-side changes are required for this, all Request and Response template expansion is performed on
+the client side.
+
+- None
+
+## Design Details
+
+**Publishing Data:**
+
+Alpha:  default container annotation
+
+- Define a well known annotation `kubectl.kubernetes.io/default-container` in a Pod to provide a way to consumers (CLI, operators) to known which is the default Container.
+- Define a global function GetDefaultContainerName for kubectl that uses this annotation like below
+```
+func GetDefaultContainerName(p *ExecOptions, pod *corev1.Pod) string{
+  containerName := p.ContainerName
+  if len(containerName) != 0 {
+    return containerName
+  }
+
+  if len(pod.Spec.Containers) > 1 {
+    // in case the "kubectl.kubernetes.io/default-container" annotation is present, we preset the opts.Containers to default to selected
+    // container. This gives users ability to preselect the most interesting container in pod.
+    if annotations := pod.Annotations; annotations != nil && len(annotations[corev1.DefaultContainerAnnotationName]) > 0 {
+      containerName := annotations[corev1.DefaultContainerAnnotationName]
+      if exists, _ := podutils.FindContainerByName(pod, containerName); exists != nil {
+        return containerName
+      } else {
+        fmt.Fprintf(p.ErrOut, "Default container name %q in annotation not found in a pod\n", containerName)
+      }
+    }
+    fmt.Fprintf(p.ErrOut, "Defaulting container name to first container %s.\n", pod.Spec.Containers[0].Name)
+
+    if p.EnableSuggestedCmdUsage {
+      fmt.Fprintf(p.ErrOut, "Use '%s describe pod/%s -n %s' to see all of the containers in this pod.\n", p.ParentCommandName, pod.Name, p.Namespace)
+    }
+  }
+
+  containerName = pod.Spec.Containers[0].Name
+  return containerName
+}
+```
```

Comment:
No need for code, let's leave something for actual implementation. You can just describe the algorithm in steps.

Reply:
Sorry, this was my fault asking for some details :)

Reply:
I will change it to some steps' description.